Replaced StaticClass() with StaticClassName() in BP classes (#379) and improved AppendString call in FName::GetRawString()

igromanru opened the pull request:
## Changes
Main focus of the PR are changes that I suggested in #379
- Added `bool IsSubclassOf(const FName& baseClassName) const;` to compare the class by its name (FName)
- Added `bool IsA(class UClass* TypeClass) const;` to use `bool IsSubclassOf(const FName& baseClassName) const;`
- Added `StaticName()` generation for all classes
- Replaced `StaticClassImpl` and `StaticBPGeneratedClassImpl` with macros
- Added functions `GetStaticName`, `GetStaticClass` and `GetStaticBPGeneratedClass` to check and assign static variables for `StaitcClass()` and `StaticName()`
- Added `nullptr` check to `GetDefaultObjImpl()` (would crash before, if `StaticClass` returned `nullptr)`
- Added `BasicFilesImpleUtils::StringToName` which converts a `wchar_t` string to a FName with help of `UKismetStringLibrary::Conv_StringToName`
- Added `#include "Engine_classes.hpp"` to **Basic.cpp**

**Additonal changes**
- Changed `BasicFilesImpleUtils::FindClassByName` to accept `bByFullName` parameter to search by full name
- Changed the `FName::GetRawString` function to use a `wchar_t` buffer and `FString`. Stack allocation is not only faster, but it should also be safer for threads in this case, since the `free` call in `FAllocatedString`’s destructor could cause a race condition.
- Added another constructor to `FString` that allows to create a FString from a pointer, `FString(wchar_t* Str, int32 Num, int32 Max)`
- Added default constructor to `FName` that will set everything to 0 per default
- Added a copy constructor to `FName`
- Added `FName::IsNone` to check if it's a `None` name
- Added assign operator to `FName`
- Added new setting `Settings::CppGenerator::XORStringInclude` that will add a new `#include` with given file name/path to the xor header file to the `Basic.hpp`
- Changed `XORString` setting comment to explain how it works

## Implementations
### StaticClass() and StaticClassImpl()
I’ve reworked `StaticClass()` to work the old-fashioned way, where it caches the static class pointer in a local static variable. No extra compiler magic needed.
```cpp
static class UClass* StaticClass()
{
	static auto ptr = StaticClassImpl("Character");
	return ptr;
}
```
`StaticClassImpl` is now basically just a wrapper that decides at compile time if `FindClassByName` or `FindClassByFullName` is used.
I forget to test it before making the PR, since it accepts the Name string as parameter now, but any xor-string should work fine with it, since it doesn't need to know the details at compile time. Compiler inlines the whole `StaticClass()` call and any xor decryption happens before. It's important to note that the parameter type is a `char*`, it's good for xor functions which usually return a char* or wchat_t* and all the magic that converts the  `char*` to a `std::string` is done by FindClassByFullName/FindClassByName. 
```cpp
template<bool bIsFullName = false>
class UClass* StaticClassImpl(const char* Name)
{
	if constexpr (bIsFullName) {
		return BasicFilesImpleUtils::FindClassByFullName(Name);
	}
	else /* default */ {
		return BasicFilesImpleUtils::FindClassByName(Name);
	}
}
```
### StaticClassName()
Since the SDK doesn’t always use GNames, and it’s generally a hassle to convert or search for a `FName` by string, I opted for a simpler solution.
```cpp
static class FName& StaticClassName()
{
	static auto className = UKismetStringLibrary::Conv_StringToName(FString(L"BP_HDPlayerCharacterBase_C"));
	return className;
}
```
**How it works and why I think it's fine:**
The compiler inlines the StaticClassName() call.  Under normal circumstances, it never gets called until you use it as a parameter in `IsA` for the first time.  
I haven’t seen any issues with it. Also, in well-written games, it’s rare that you need to check for the topmost class, you can usually use `IsA` on the `StaticClass` of a base class.  
Xor string will work fine here as well.  
### BasicFilesImpleUtils::FindClassByName changes
I changed `FindClassByName` initially because I wanted to use it instead of `StaticClassImpl` but later decided against it.
I still find the change useful, so I left it as is.

## Postscript
The PR is open for review but shouldn’t be merged yet.
I doubt it will be approved right away, but I’m still testing it and may make some changes.  
Besides that, I would appreciate any feedback ASAP so I can work on it while it’s still fresh in my memory.

Merge of the PR will close #379

Fischsalat commented:
I know it's a draft but you should probably get it to compile first.
<img width="715" height="358" alt="image" src="https://github.com/user-attachments/assets/f0e3a81c-f0cc-4887-84e5-94e991f04015" />

I don't want to need to point out every error that could cause the SDK not to compile, when the compiler does it just the same.

Fischsalat commented:
> I'm using SDK generated by the Fork in two projects now. It compiled fine, but I also haven't used GetDefaultObj() on a Blueprint class to notice it.

You don't need to call the function, you just need to include any file with a blueprint generated class. Always include SDK.hpp when making changes to the generated SDK, else it wont compile the majority of the SDK.

> We have multiple options here, like to keep the old logic, return a nullptr or do something else

Returning a nullptr would probably not be the best option, as GetDefaultObj does `StaticClass()->DefaultObject`.
(Note: I've just realized that GetDefaultObj() will always crash when called at a time, at which the class isn't loaded. Not good.).

Perhaps one solution is to change the access of `StaticClass()` to private. This way the current `GetDefaultObj` implementation could be kept (or at least a version with a fix for the above note), while still forcing the user to use `StaticClassName()` for type-checking.

> do Blueprints actually have default objects?
Yes.

> Pascal Case for local variable names and no auto right?
Yes.

igromanru commented:
> > I'm using SDK generated by the Fork in two projects now. It compiled fine, but I also haven't used GetDefaultObj() on a Blueprint class to notice it.
> 
> You don't need to call the function, you just need to include any file with a blueprint generated class. Always include SDK.hpp when making changes to the generated SDK, else it wont compile the majority of the SDK.

Oh, you mean the generated SDK.hpp? I was using my own with selected includes, but few of them were BP classes.  
But yeah, makes sense to test with full SDK.

> > We have multiple options here, like to keep the old logic, return a nullptr or do something else
> 
> Returning a nullptr would probably not be the best option, as GetDefaultObj does `StaticClass()->DefaultObject`.
> 
> Perhaps one solution is to change the access of `StaticClass()` to private. This way the current `GetDefaultObj` implementation could be kept (or at least a version with a fix for the above note), while still forcing the user to use `StaticClassName()` for type-checking.

Alright, I'll revert to generate `StaticClass()` for BP as well and try to make them private.

>  (Note: I've just realized that GetDefaultObj() will always crash when called at a time, at which the class isn't loaded. Not good.).

It's possible to use `LoadAsset` function to load an object into memory, but it's very slow. No idea what would be a good way to fix `GetDefaultObj()` for BPs.  

> Also what do you think about renaming `StaticClassName` to just `StaticName`?

If I think about it, the most accurate name would be `ClassName`. Since it's the name you get from the `Class` property (`this->Class->Name`). It's not static like the StaticClass, it's just the short name of the class, which is usually always unique.

Fischsalat commented:
Sorry for the retarded commit that I somehow appended to your pullrequst, github desktop is extremely retarded and I'm just trying to pull the latest changes on this pr.

igromanru commented:
Np, I had to force push anyway.  

So I tried to make `StaticClass()` function private in BP classes, and it didn't work out.  
If it's private, `GetDefaultObjImpl()` can't use it and `GetDefaultObjImpl()` is needed as a compiler trick, since UClass is only forward declarated in some files.  
Then it occurred to me: why don’t we just leave `StaticClass()` for Blueprints and let SDK users decide for themselves whether they want to use it? 
It might be a good idea to add a note in the UsingTheSDK guide that `ClassName()` should be used for BP classes.  
However, technically, there could be rare cases where someone needs the `StaticClass` of a BP.

I'm still working on the code and will test it further tomorrow.  
I'll let you know when you can review it again.

Fischsalat commented:
Hm, yea, I guess if you want to spawn an object for example. You can't really pass the class as a parameter if StaticClass is private.

Maybe as a worst case solution we could force users not to use the result of a StaticClass call on a blueprint-class by wrapping it in a template type that is not convertible to a different type, that's used as a wrapper the Class parameter in IsA.

```
TBlueprintClass<ClassType> StaticClass();

bool IsA(ComparisonClass Class); 
```
Where there is no conversion between
`TBlueprintClass` and `ComparisonClass`.

igromanru commented:
> Maybe as a worst case solution we could force users not to use the result of a StaticClass call on a blueprint-class by wrapping it in a template type that is not convertible to a different type, that's used as a wrapper the Class parameter in IsA.

Currently, I’m in favour of just leaving it as is. In the meantime, I don’t think there’s a need to hold the SDK user’s hand on this one.  
I've simplified `StaticBPGeneratedClassImpl` to receive `const char* Name` as parameter, and now it works with XOR string as well.  
And I added a simple `nullptr` check to `GetDefaultObjImpl`, so it no longer crashes.

My latest changes are ready for review. I've compiled full SDK for Operation: Harsh Doorstop and Remnant 2 with and without XOR string and tested simple hack with Player ESP and Radar in Operation: Harsh Doorstop.  
Tested `StaticName()` and `GetDefaultObj()` on BP classes that were both in memory and not.

I've run into problems that `Engine_classes.hpp` and `CoreUObject_classes.hpp` aren't included in some files (which is normal).  
Like I mentioned before, I couldn't replace `GetDefaultObjImpl` because of missing `CoreUObject_classes.hpp` include.  

But `Engine_classes.hpp` caused the problem that `KismetStringLibrary` isn’t known.  
Here comes the important part that I need your feedback on:  
My solution to deal with it was to recreate `KismetStringLibrary::Conv_StringToName` and place it in the `BasicFilesImpleUtils` namespace.  
From what I’ve seen, the signature of `Conv_StringToName` has never changed in any UE4 or UE5 versions, so I assume it shouldn’t cause any problems.  
Additionally, keeping it as its own function allows the logic to be replaced later with something else if needed.

I also have edited UsingTheSDK.md a bit to provide a bit more info which `IsA` should be used.

igromanru commented:
> > My solution to deal with it was to recreate KismetStringLibrary::Conv_StringToName and place it in the BasicFilesImpleUtils namespace.
> 
> Well, now you know why StaticClassImpl exists. Please just create one function for the `StaticName` implementation.

I'm unsure what you mean by that. Do you want me to make `BasicFilesImpleUtils::StringToName` into `StaticNameimpl` instead?  

> > I don’t think there’s a need to hold the SDK user’s hand on this one.
> 
> There's almost always a need to do this, you don't understand how incredibly lost some of the people are.

I know, but that doesn’t mean we have to. Currently, all these functions are safe to use, they cause no crashes, and in the worst-case scenario, they might only slow down the game or not work properly for IsA.  
On the other hand, if we try to gatekeep them, it might restrict people who know what they’re doing from using them in legitimate cases.

Fischsalat commented on Dumper/Generator/Private/Generators/CppGenerator.cpp:
Instead of copying `UKismetSystemLibrary::Conv_StringToName` you should just call the function.

igromanru replied:
You have to teach me how. Because the compiler don't know what `UKismetStringLibrary` class is and like I mentioned before, some files don't even include `Engine_classes.hpp`. 
I couldn't make it like `GetDefaultObjImpl` where you call `ClassType::StaticClass()` because I can't pass `UKismetStringLibrary` as ClassType like it happens with classes.  
The compiler don't know about  `UKismetStringLibrary` either in `Basic.hpp` nor in packages like `CoreUObject` (CoreUObject_classes.hpp) which only includes the `Basic.hpp`.

Fischsalat commented:
Why does StaticName return a non-const reference to the static variable?


The current StaticClass implementation wont work for people who are too stupid to properly manual map their DLLs. Some people are challenged and can not properly initialize static variables. This alone I wouldn't care about but they then always open issues on this repo and waste my time. So the StaticClass implementation would need to be replaced by this:
```c++
static class UClass* StaticClass()
{
	static UClass* Clss = nullptr;
		
	if (Clss == nullptr)
		Clss = GetStaticBPGeneratedClass("BP_Buildable_Ghost_Preview_C");

	return Clss;
}
```
I'm would strongly prefer to delegate initialization to an implementation-funtion, rather than to extend the function by 4 more lines.
No, compressing the code like in your current `GetDefaultObjImpl` is not a solution and just looks ugly.

igromanru commented:
Replaced `StaticClass()` and `StaticName()` implementations with the macros you suggested.

Fischsalat reviewed (commented):
Look pretty good, did you test the SDK on several different games already? Not just if it compiles but also whether StaticName, etc. work as intended.
